File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/H._Числовые_пути/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(int value, Node* left, Node* right) : value(value), left(left), right(right) {}
};
#endif

#ifdef REMOTE_JUDGE
#include "solution_tree.h"
#endif
#include <cmath>
#include <iostream>
#include <string>
#include <cassert>

using namespace std;


void dfs(const Node* root, std::string& path, int& totalSum) {
    path += std::to_string(root->value);
    if (!root->left && !root->right) totalSum += std::stoi(path);
    if (root->left) dfs(root->left, path, totalSum);
    if (root->right) dfs(root->right, path, totalSum);
    path.pop_back();
}


int Solution(const Node* root) {
    /*
    Time Complexity O(N)
    Memory Complexity O(H)
    */
    int totalSum = 0;
    std::string path = "";
    if (root) dfs(root, path, totalSum);
    return totalSum;
}

#ifndef REMOTE_JUDGE
void test() {
    Node node1({2, nullptr, nullptr});
    Node node2({1, nullptr, nullptr});
    Node node3({3, &node1, &node2});
    Node node4({2, nullptr, nullptr});
    Node node5({1, &node4, &node3});
    assert(Solution(&node5) == 275);
}

int main() {
  test();
}
#endif

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/H._Числовые_пути/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, value, left=None, right=None):
            self.value = value
            self.right = right
            self.left = left


def solution(root: 'Node') -> int:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    total_sum: int = 0
    path: list[str] = list()
    def dfs(root: 'Node', path: list[str]) -> None:
        path.append(str(root.value))
        if not root.left and not root.right:
            nonlocal total_sum
            total_sum += int(''.join(path))
        if root.left: dfs(root.left, path)
        if root.right: dfs(root.right, path)
        path.pop()
    if root: dfs(root, path)
    return total_sum


def test():
    node1 = Node(2, None, None)
    node2 = Node(1, None, None)
    node3 = Node(3, node1, node2)
    node4 = Node(2, None, None)
    node5 = Node(1, node4, node3)

    assert solution(node5) == 275


if __name__ == '__main__':
    test()